ros/src/twist_controller/twist_controller_mod.py

```diff
@@ -32,60 +32,55 @@ def __init__(self, default_update_interval, wheel_base, steer_ratio, min_speed,
         self.setup_pid_controllers(config['dyn_velo_proportional_control'], config[
-            'dyn_accel_proportional_control'], config['dyn_accel_integral_control'], self.max_deceleration, self.max_acceleration)
-
-        # reset acceleration cruise constant
-        self.acceleration_cruise_constant = config['dyn_acceleration_cruise_constant']
+            'dyn_velo_integral_control'], config['dyn_braking_proportional_control'], config['dyn_braking_integral_control'])
 
         return config
 
-    def setup_pid_controllers(self, velo_p, accel_p, accel_i, max_deceleration, max_acceleration):
-        rospy.loginfo("Initializing PID controllers with velo_P: {}, accel_P: {}, accel_I: {}"
-                      .format(velo_p, accel_p, accel_i))
+    def setup_pid_controllers(self, velo_p, velo_i, braking_p, braking_i):
+        rospy.loginfo("Initializing PID controllers with velo_P: {}, velo_I: {}, braking_P: {}, braking_I: {}"
+                      .format(velo_p, velo_i, braking_p, braking_i))
         # create velocity pid controller thresholded between min and max
         # acceleration values
         self.velocity_pid_controller = PID(
-            velo_p, 0, 0, max_deceleration, max_acceleration)
+            velo_p, velo_i, 0, 0, 1)
 
         # create acceleration pid controller thresholded between 0% and 100%
         # for throttle
-        self.acceleration_pid_controller = PID(accel_p, accel_i, 0.0, 0.0, 1.0)
+        self.braking_pid_controller = PID(
+            braking_p, braking_i, 0.0, 0.0, 10000)
 
     def control(self, target_linear_velocity, target_angular_velocity, current_linear_velocity):
         # compute timestep
         timestep = self.compute_timestep()
-        # use velocity pid controller to compute desired acceleration
-        acceleration_command = self.compute_acceleration_command(
-            target_linear_velocity - current_linear_velocity, timestep)
-        rospy.logwarn('Current linear velocity %.2f, target linear velocity %.2f, desired total acceleration %.2f',
-                      current_linear_velocity, target_linear_velocity, acceleration_command)
+        velocity_error = target_linear_velocity - current_linear_velocity
+
+        if (target_angular_velocity == 0 and current_linear_velocity == 0):
+            # reset integrators if we're at a stop
+            self.reset()
+
+        # use throttle if we want to speed up or if we want to slow down just slightly
+        throttle_command = self.velocity_pid_controller.step(
+            velocity_error, timestep) if velocity_error > 0 or velocity_error > (-0.075 * current_linear_velocity) else 0
+
+        # use brake if we want to slow down somewhat significantly or it looks like want to stop
+        brake_command = self.braking_pid_controller.step(-velocity_error, timestep) if velocity_error < (-0.1 * current_linear_velocity) or (velocity_error < 0 and current_linear_velocity < 2.5) else 0
+
```

Comment:
For the braking threshold approaching a stop, this condition keeps the brake PID active but as the car's speed approaches the target speed of 0, the PID brake command will also lighten up back to zero so we still need to trigger some manual holding of a fixed brake amount to keep the car stopped.  The simulator car doesn't have any creep torque so it can stay stopped with throttle=0 and brake=0 but a real car will creep forward.  Also, the waypoint updater target speed at very low speeds can't be smooth because it's set by the waypoint spacings, so we should probably just switch over to manual brake control anyways when speed < 1~1.5 m/s.

The simplest way to do this for example is if speed < 1.5 m/s and throttle command == 0, then set a fixed brake command = 50Nm.  Another way would be to detect a slow down condition like speed crossing from >1.5 m/s to <1.5 m/s and throttle==0, then latch the previous value of PID brake control to keep it fixed.

Reply:
Do we know for sure that Carla behaves this way? Certainly, I've never seen a car that doesn't have an idle acceleration (even pure electric vehicles), but I just want to make sure that I'm not coding for a situation that doesn't exist. If we're that the simulator just doesn't simulate idle acceleration, then I'll add that in shortly.

Reply:
Yea I’m 99% sure the Lincoln MKZ hybrid has simulated traditional creep torque.  I think the data from the Udacity ROS bag also shows steady brake apply while the car is stopped at the light, but I can double check again later.

Reply:
Don't worry about double checking; I'll take your word for it and add that now.